### PR TITLE
Add sign to performance test graphs

### PR DESCRIPTION
**Pull request by blindpirate**
### Context

This PR did some improvements for https://github.com/gradle/gradle-private/issues/1685 . Previously we display confidence and difference in different graphs, which make it hard to recognize whether a high confidence result is improvement or regression. This PR adds a background color for the confidence graph:

image 

Red/Orange is regression and Green/LightGreen is improvement.

I also want to keep the difference graph because we can easily see fluctuation and trend from it.

### CODE REVIEW

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/TestPageGenerator.java [hunk 1]
```diff
@@ -187,10 +189,10 @@ private PerformanceTestExecution findPreviousExecutionInSameBranch(PerformanceTe
             }
 
             private void addPerformanceGraphs() {
-                List<Chart> charts = Lists.newArrayList(new Chart("totalTime", "total time", "s", "totalTimeChart"));
+                List<Chart> charts = Lists.newArrayList(new Chart("totalTime", "total time", "s", "totalTimeChart", false));
                 if(testHistory instanceof CrossVersionPerformanceTestHistory) {
-                    charts.add(new Chart("confidence", "confidence", "%", "confidenceChart"));
-                    charts.add(new Chart("difference", "difference", "%", "differenceChart"));
+                    charts.add(new Chart("confidence", "confidence", "%", "confidenceChart", true));
+                    charts.add(new Chart("difference", "difference", "%", "differenceChart", false));
```
Comment (wolfs) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/TestPageGenerator.java]: I was expecting that we add the background to the difference graph and the sign to the confidence graph.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/TestDataGenerator.java [hunk 2]
```diff
@@ -163,4 +173,37 @@ public String getDate() {
...
+        private static final String LIGHT_GREEN = "#80ff00";
+        private static final String GREEN = "#00ff00";
+        private Map xaxis;
+        private String color;
+
+        private BackgroundColor(List<Number> xy) {
+            double index = xy.get(0).doubleValue();
+            double differencePercentage = xy.get(1).doubleValue();
+            xaxis = ImmutableMap.of("from", index - 0.5, "to", index + 0.5);
+            if (differencePercentage > 2) {
```
Comment (wolfs) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/TestDataGenerator.java]: If we use confidence here, I would say that anything < 80% confidence is a white background. Then we could scale from light red to darker red (probably with some opacity) evenly between 80-100%. We would do the same with "negative" confidence, i.e. an improvement, with green.

**Comment by blindpirate** [comment]
Here's what it looks like right now:

@wolfs Can you please take a look again?

image

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/TestDataGenerator.java [hunk 3]
```diff
@@ -163,4 +173,43 @@ public String getDate() {
...
+        private static final String LIGHT_GREEN = "#80ff00";
+        private static final String GREEN = "#00ff00";
+        private Map xaxis;
+        private String color;
+
+        private static BackgroundColor ofConfidence(List<Number> xy) {
+            double index = xy.get(0).doubleValue();
+            double confidencePercentage = xy.get(1).doubleValue();
+            if (confidencePercentage > 90) {
+                return new BackgroundColor(index, RED);
```
Comment (wolfs) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/TestDataGenerator.java]: I would have preferred to use a linear background color, so something like
```
int limit = 80
int aboveLimit = confidencePercentage - limit;
int redValue = aboveLimit * 256 / (100 - limit);
```
And then use the `redValue` for the red component of the background. Same goes for the green value for improvements.

We can also start out with what you have right now and improve later on.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/TestDataGenerator.java [hunk 4]
```diff
@@ -175,33 +175,33 @@ public String getDate() {
...
             } else {
-                return new BackgroundColor(index, GREEN);
+                return null;
             }
         }
 
-        private BackgroundColor(double index, String color) {
+        // See TestDataGeneratorTest for examples
+        private static int redComponent(double confidencePercentage) {
+            return (int) (128 + (THRESHOLD + confidencePercentage) * 128 / (100 - THRESHOLD));
```
Comment (wolfs) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/TestDataGenerator.java]: Interesting formula. I think I finally understand what is going on here - it is hard though...

**Comment by blindpirate** [comment]
Thanks @wolfs !

**Pull request merged**